File: Year 1/Semester 2/Java/Lessons/interfaces.java
public class interfaces {
    
    // Define an interface
public interface Animal {
    void makeSound();
    void move();
}

// Implement the interface in a class
public class Dog implements Animal {
    @Override
    public void makeSound() {
        System.out.println("Woof!");
    }

    @Override
    public void move() {
        System.out.println("The dog runs.");
    }
}

// Another implementation of the interface
public class Cat implements Animal {
    @Override
    public void makeSound() {
        System.out.println("Meow!");
    }

    @Override
    public void move() {
        System.out.println("The cat sneaks.");
    }
}

public class Main {
    public static void main(String[] args) {
        // Creating instances of Dog and Cat
        Animal dog = new Dog();
        Animal cat = new Cat();

        // Calling interface methods
        dog.makeSound();
        dog.move();
        
        cat.makeSound();
        cat.move();
    }
}

    
}

// Why And When To Use Interfaces?

// 1) To achieve security - hide certain details and only show the important details of an object (interface).

// 2) Java does not support "multiple inheritance" (a class can only inherit from one superclass).
//  However, it can be achieved with interfaces, because the class an implement multiple interfaces. 
// Note: To implement multiple interfaces, separate them with a comma.

File: Year 1/Semester 1/C/Lessons/interfaces.c
#include <stdio.h>

// Define a struct for Animal with function pointers
typedef struct Animal {
    void (*makeSound)(); // Function pointer for makeSound
    void (*move)();      // Function pointer for move
} Animal;

// Implementation for Dog
void dogMakeSound() {
    printf("Woof!\n");
}

void dogMove() {
    printf("The dog runs.\n");
}

// Implementation for Cat
void catMakeSound() {
    printf("Meow!\n");
}

void catMove() {
    printf("The cat sneaks.\n");
}

// Main function to demonstrate the use of interfaces
int main() {
    // Creating instances of Dog and Cat
    Animal dog = { dogMakeSound, dogMove }; // Initialize Dog
    Animal cat = { catMakeSound, catMove }; // Initialize Cat

    // Calling interface methods
    dog.makeSound(); // Call Dog's makeSound
    dog.move();      // Call Dog's move

    cat.makeSound(); // Call Cat's makeSound
    cat.move();      // Call Cat's move

    return 0;
}
